### ISSUE TITLE
Grid kernel

### ISSUE THREAD

**Issue by sokratiag** [cornellius-gp/gpytorch#1278]
I have some spatial data (4990 locations) which lies on a non-regular 2D grid. Is it possible to use the GridKernel to speed up  computations? In this case, train_x is given as (x_i, y_i) locations, could the grid argument be the same as train_x?

**Comment by gpleiss** [comment]
What do you mean by a non-regular 2D grid. Chances are GridKernel won't work, but GridInterpolationKernel should be nearly just as good with the same computational complexity.

**Comment by sokratiag** [comment]
The grid is not uniform - I tried the GridInterpolationKernel and it does speed up the computations. Should my data live on a uniform grid in order to see speed ups with the gridKernel? Thanks!

**Comment by gpleiss** [comment]
What kind of numbers are you seeing? How big of a grid size did you use with GridInterpolationKernel? A code example could help debug this.

With N \approx 5000 you should see some speed up.

**Comment by sokratiag** [comment]
For the gridInterpolationKernel I used the utility gpytorch.utils.grid.choose_grid_size which gave 70 points and it works fine in terms of speeding the process up. 

My problem is that I can not see speed ups with the gridKernel despite that my inputs are on a grid. Here is an example where data is read from files. The train_x is spatial data arranged on a non-uniform 2D grid  (x,y)(4992 locations)

Y = pd.DataFrame(B)
X = locations[['x', 'y']]
Y =Y.T
Y = Y.loc[X.index]

train_y = Y
train_x = torch.Tensor(train_x[::])

grid = grid_loc[['xgrid', 'ygrid']]
grid = np.array(grid)
grid = torch.Tensor(grid[::])

class GridGPRegressionModel(gpytorch.models.ExactGP):
        def __init__(self, grid, train_x, train_y, spatial_likelihood):
            super(GridGPRegressionModel, self).__init__(train_x, train_y, spatial_likelihood)
            num_dims = train_x.size(-1)
            
            self.mean_module = gpytorch.means.ConstantMean()

self.covar_module = gpytorch.kernels.GridKernel(
            gpytorch.kernels.ScaleKernel(
                 gpytorch.kernels.RBFKernel()
             ), 
            grid=grid)

 def forward(self, x):
            mean_x = self.mean_module(x)
            covar_x = self.covar_module(x)
            return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

    spatial_likelihood = gpytorch.likelihoods.GaussianLikelihood()

    
    spatial_model = GridGPRegressionModel(grid, train_x, train_y, spatial_likelihood)

spatial_likelihood.train()
    
    optimizer = torch.optim.LBFGS(spatial_model.parameters(), lr=1, max_iter=1000,line_search_fn='strong_wolfe')


    # # "Loss" for GPs - the marginal log likelihood
    spatial_mll = gpytorch.mlls.ExactMarginalLogLikelihood(spatial_likelihood, spatial_model)
    
    def closure():
        optimizer.zero_grad()
        output = spatial_model(train_x)
        spatial_loss = -spatial_mll(output, train_y)
        spatial_loss.backward()
        marginal_likelihood = spatial_mll(output,torch.Tensor(train_y))
        print('Loss: %.3f' % ( spatial_loss.item()))
        print('gpytorch ML {}'.format(torch.sum(marginal_likelihood)))
     

        return spatial_loss 
   
    optimizer.step(closure)


Thank you for your help.

**Comment by gpleiss** [comment]
Take a look at the [Grid Structure Regression Example](https://docs.gpytorch.ai/en/v1.2.0/examples/02_Scalable_Exact_GPs/Grid_GP_Regression.html). `train_x` and `grid` have to correspond exactly.

Note this line in the example:
```python
train_x = gpytorch.utils.grid.create_data_from_grid(grid)
```

**Comment by gpleiss** [comment]
> The train_x is spatial data arranged on a non-uniform 2D grid (x,y)(4992 locations)

If the grid is non-uniform, then `GridKernel` won't work and you'll have to use `GridInterpolationKernel`.
